core/domain/learner_progress_services.py

```diff
@@ -0,0 +1,231 @@
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS-IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Services for tracking the progress of the learner."""
+
+from core.platform import models
+from core.domain import subscription_services
+
+(user_models,) = models.Registry.import_models([models.NAMES.user])
+
+def add_exp_id_to_completed_list(user_id, exp_id):
+    """Adds the exploration id to the completed list of the user.
+
+    Callers of this function must ensure that the user id and the
+    exploration id are valid.
+    """
+    activities_completed_model = (
+        user_models.ActivitiesCompletedByLearnerModel.get(
+            user_id, strict=False))
+    if not activities_completed_model:
+        activities_completed_model = (
+            user_models.ActivitiesCompletedByLearnerModel(id=user_id))
+
+    exploration_created_ids = (
+        subscription_services.get_exploration_ids_subscribed_to(user_id))
+
+    if (exp_id not in exploration_created_ids and
+            exp_id not in activities_completed_model.completed_exp_ids):
+        # Remove the exploration from the in progress list (if present) as it is
+        # now completed.
+        remove_exp_from_incomplete_list(user_id, exp_id)
+        activities_completed_model.completed_exp_ids.append(exp_id)
+        activities_completed_model.put()
+
+def add_collection_id_to_completed_list(user_id, collection_id):
+    """Adds the collection id to the list of collections completed by the user.
+
+    Callers of this function must ensure that the user id and collection id are
+    valid.
+    """
+    activities_completed_model = (
+        user_models.ActivitiesCompletedByLearnerModel.get(
+            user_id, strict=False))
+    if not activities_completed_model:
+        activities_completed_model = (
+            user_models.ActivitiesCompletedByLearnerModel(id=user_id))
+
+    collection_created_ids = (
+        subscription_services.get_collection_ids_subscribed_to(user_id))
+
+    if (collection_id not in collection_created_ids and
+            collection_id not in
+            activities_completed_model.completed_collection_ids):
+        # Remove the collection from the in progress list (if present) as it is
+        # now completed.
+        remove_collection_from_incomplete_list(user_id, collection_id)
+        activities_completed_model.completed_collection_ids.append(
+            collection_id)
+        activities_completed_model.put()
+
+def add_exp_to_incomplete_list(
+        user_id, exploration_id, timestamp, state_name, version):
+    """Adds the exploration id to the incomplete list of the user.
+
+    Callers of this function must ensure that all the arguments provided are
```

Comment:
I just included this as a safeguard but you are right - it has no real purpose. Removed it.